# TD 4: Listes

---

## 1. Débuts avec les listes

Dans cette partie, il faut écrire les réponses directement dans le notebook. En plus du cours, n'hésitez pas à consulter la documentation (tutoriel) Python sur les listes [ici (introduction aux listes)](https://docs.python.org/fr/3.6/tutorial/introduction.html#lists) et [ici (complément sur les listes)] (https://docs.python.org/fr/3/tutorial/datastructures.html#more-on-lists).

1. Créer la liste `3, 5, 10`, l'affecter à une variable, puis l'afficher.

In [15]:
liste= [3,5,10]
print(liste)

[3, 5, 10]


2. Ajouter ensuite 12 et 17 à la fin puis afficher à nouveau la liste.

In [16]:
liste.extend([12, 17])

print(liste)

[3, 5, 10, 12, 17]


3. Changer le 10 en -7.

In [20]:
liste[2]=-7
print(liste)

[3, 5, -7, 12, 17]


In [17]:
liste[liste.index(10)]=-7
print(liste)

[3, 5, -7, 12, 17]


4. Avec une boucle `for`, écrire le double de chaque élément de la liste. Cela doit afficher `[6, 10, -14, 24, 34]`. La liste doit être modifiée.

In [18]:
for i in range(len(liste)):
    liste[i]*=2
print(liste)

[6, 10, -14, 24, 34]


5. A l'aide d'une boucle `for`, ajouter  la valeur `i` à l'élément d'indice `i` de la liste. Cela doit afficher `[6, 11, -12, 27, 38]`. La liste doit être modifiée.

In [19]:
for i in range(len(liste)):
    liste[i]+=i
print(liste)


[6, 11, -12, 27, 38]


6. Tirer 10 nombres à 2 chiffres au hasard et les mettre dans une nouvelle liste (vide au départ), puis afficher la somme des entiers de la liste et le plus grand de ces entiers.

In [20]:
import random
liste_1 = []
for i in range(10):
    liste_1.append(random.randint(10,99)) #ecrivable liste_1=[random.randint(10.99) for i in range(10)]
print(sum(liste_1))
print(max(liste_1))

554
99


In [61]:
import random
liste_a= []
liste_a=[random.randint(10,99) for i in range(10)]
print(sum(liste_a))
print(max(liste_a))
print(liste_a)

514
96
[32, 65, 67, 14, 50, 19, 96, 36, 77, 58]


7. A partir de la liste choisie au hasard, créer une liste qui contient les éléments pairs et une deuxième liste qui contient les éléments impairs.

In [23]:
liste_paire = []
liste_impaire = []
for element in liste_1:
    if element%2 == 0:
        liste_paire.append(element)
    else:
        liste_impaire.append(element)
print(liste_paire)
print(liste_impaire)

[26, 30, 46, 80, 96]
[55, 55, 55, 73, 97]


8. Trier la liste choisie au hasard par ordre croissant en modifiant la liste.

In [21]:
liste_1.sort()
print(liste_1)

[12, 31, 41, 45, 49, 55, 58, 75, 89, 99]


9. Supprimer le plus petit et le plus grand élément de la liste triée.

In [22]:
liste_1.remove(liste_1[0]) #équivaut à : del liste_1[0] ou min(liste_1)
liste_1.remove(liste_1[-1])
print(liste_1)

[31, 41, 45, 49, 55, 58, 75, 89]


10. Créer la liste des entiers de 1 jusqu'à 52. Ces entiers représentent les cartes d'un paquet de 52 cartes.

In [53]:
liste_ent = []   #équivaut à paquet=[ i for i in range (1,53)] ou paquet=list(range(1,53))
for i in range(1,53):
    liste_ent.append(i)
print(liste_ent)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52]


In [47]:
paquet=list(range(1,53))
print(paquet)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52]


In [46]:
paquet=[ i for i in range (1,53)]
print(paquet)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52]


 11. Couper le jeu de carte revient à choisir un indice `i` entre 0 et 51, et à passer toutes les cartes allant des indices 0 à i-1 à la suite des cartes qui vont des indices i à 51. Ecrire le code qui effectue une coupe choisie au hasard.

In [29]:
import random
i= random.randint(0,51)
for n in range(i):
    liste_ent.pop(0)
    liste_ent.append(n+1)
print(liste_ent)
print(i)

[20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
19


In [54]:
import random
coupe= random.randint(1,53)
paquet_z=[]                         #= liste_ent=(liste_ent+liste_ent[0:coupe])[coupe:]
for i in range(coupe,52):
    paquet_z.append(liste_ent[i])
for i in range(coupe):
    paquet_z.append(liste_ent[i])
print(paquet_z)
print(coupe)

[42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41]
41


In [ ]:
#correction
paquet_z=[]                         #= liste_ent=(liste_ent+liste_ent[0:coupe])[coupe:]
for i in range(coupe,52):
    paquet_z.append(liste_ent[i])
for i in range(coupe):
    paquet_z.append(liste_ent[i])
print(paquet_z)

#correction2
liste_ent=(2*liste_ent)[coupe:coupe+52]
print(liste_ent)

---

## 2. Problème de Syracuse

Cette partie est à traiter dans un script que vous pouvez appeler, par exemple, `syracuse.py`.

En partant d'un entier `n` de départ, on définit une suite d'entiers en obtenant chaque nouveau terme à partir du précédent soit en le divisant par 2 s'il est pair, soit en le multipliant par 3 et en ajoutant 1 s'il est impair. 

1. Sans écrire de programme, calculer les premières valeurs de la suite en chosissant `n = 3`.

2. Ecrire la fonction qui, à partir d'un entier initial, renvoie la liste des valeurs successives jusqu'à ce que la valeur `1` soit atteinte (le dernier élément de la liste est donc toujours 1).

In [35]:
def syracuse(n):
    """ Retourne la liste des valeurs de la suite en partant de n jusqu'à 1 """
    l=[n]
    while n!=1:
        if n%2 == 0:
            n //=2
        else:
            n=(n*3+1)
        l.append(n)
    return l

print(syracuse(3))

[3, 10, 5, 16, 8, 4, 2, 1]


3. La conjecture de Collatz (ou problème de Syracuse) affirme que, quel que soit l’entier `n` que l’on choisisse au départ, on finit toujours par arriver à 1 (ce résultat est une *conjecture*, c'est-à-dire qu'il n'a pas été démontré, mais qu'il n'existe pas de contre-exemple connu). Écrire une fonction qui, en appelant la fonction précédente, va vérifier si la conjecture est vraie pour `n` de 1 à `n_max`, où `n_max` est un paramètre de la fonction. 

*Remarque*: si tout se passe bien, la fonction doit juste se terminer et renvoyer `True` par exemple. Sinon, c’est qu’on sera entré dans une boucle infinie.

In [56]:
def testeConjecture(n_max):
    """ Teste la conjecture de Collatz pour toutes les valeurs de 1 à n_max """
    for i in range(1,n_max+1):
        syracuse(i)
    return True

print(testeConjecture(10000))


True


4. On appelle *temps de vol* de l’entier `n` le nombre d’étapes pour aller de `n` jusqu’à 1. Le temps de vol de 1 est 0, le temps de vol de 3 est 7. Écrire une fonction qui, étant donné un paramètre `n`, renvoie son temps de vol.

In [39]:
def temps_vol(n):
    """ Retourne le temps de vol de n """
    return len(syracuse(n))-1

print("Le temps de vol de", 3, "est", temps_vol(3))

Le temps de vol de 3 est 7


5. Ecrire une fonction qui, étant donné un paramètre `n_max` renvoie la liste des temps de vol de tous les entiers de 1 à `n_max`. *Indication*: utiliser une liste en compréhension.

In [42]:
def tempsVolListe(n_max):
    """ Retourne la liste de tous les temps de vol de 1 à n_max """
    return [temps_vol(i) for i in range(1, n_max + 1)]

l= tempsVolListe(10000)
t_max= max(l)
print("l'entier", l.index(t_max)+1, "a le plus grand temps", t_max)


l'entier 6171 a le plus grand temps 261


6. Déterminer quel entier entre 1 et 10000 a le plus grand temps de vol (réponse 6171 qui a le temps de vol 261).

7. L’altitude maximale de l'entier `n` est la plus grande valeur par laquelle passe `n` au cours de son vol. Déterminer quel entier entre 1 et 10000 a la plus grande altitude maximale (réponse 27114424, atteint par l'entier 9663). Ne pas hésiter à écrire de nouvelles fonctions pour cela.

In [ ]:
def alt_max(n):
    return max(syracuse(n))
def alt_max_liste(alt_max):
    return [alt_max(i) for i in range(1, alt_max + 1)]
l_alt= alt_max_liste(10000)
b=max(l_alt)
print("l'entier")

---

## 3. Carré magique

*Définition* extraite de [l'article Wikipedia sur le carré magique](https://fr.wikipedia.org/wiki/Carr%C3%A9_magique_(math%C3%A9matiques)): en mathématiques, un carré magique d’ordre $n$ est composé de $n^2$ entiers strictement positifs, écrits sous la forme d’un tableau carré. Ces nombres sont disposés de sorte que leurs sommes sur chaque rangée, sur chaque colonne et sur chaque diagonale principale soient égales. On nomme alors *constante magique* la valeur de ces sommes. 

Un exemple de carré magique:

$$\begin{bmatrix} 4 & 14 & 15 & 1 \\ 9 & 7 & 6 & 12 \\ 5 & 11 & 10 & 8 \\ 16 & 2 & 3 & 13 \end{bmatrix}$$

1. Sans écrire de programme, vérifier qu'il s'agit bien d'un carré magique, et donner sa constante magique. 34

Les questions qui suivent doivent être traitées dans un script, qui peut s'appeler `carre_magique.py`. Pensez à utiliser [PythonTutor](http://pythontutor.com/) et les outils de programmation pour résoudre les problèmes qui pourraient survenir.

2. Créer une liste à deux dimensions affectée à  la variable `carre_mag` contenant ce carré magique.

In [92]:
carre_mag=[[4,14,15,1],[9,7,6,12],[5,11,10,8],[16,2,3,13]]
for i in range(len(carre_mag)):
    for j in range(len(carre_mag[i])):
        print(carre_mag[i][j], end=" ")
    print()

4 14 15 1 
9 7 6 12 
5 11 10 8 
16 2 3 13 


3. Créer une deuxième liste affectée à la variable `carre_pas_mag` qui ne soit pas magique à partir du carré magique en changeant le 3 en 7. 

In [93]:
carre_pas_mag= carre_mag
carre_pas_mag[3][2]=7
print(carre_pas_mag)

[[4, 14, 15, 1], [9, 7, 6, 12], [5, 11, 10, 8], [16, 2, 7, 13]]


4. Créer une fonction qui affiche la liste comme un carré.

In [100]:
def afficheCarre(carre):
    """ Affiche la liste à 2 dimensions carre comme un carré"""
    for i in range(len(carre)):
        for j in range(len(carre[i])):
                print(carre[i][j], end=" ")
        print()
carre_mag=[[4,14,15,1],[9,7,6,12],[5,11,10,8],[16,2,3,13]]
afficheCarre(carre_mag)
afficheCarre(carre_pas_mag)


4 14 15 1 
9 7 6 12 
5 11 10 8 
16 2 3 13 
4 14 15 1 
9 7 6 12 
5 11 10 8 
16 2 7 13 


5. Ecrire une fonction qui teste si les lignes du carré ont toutes la même somme. Cette fonction renvoie cette somme si c'est le cas, et `-1` sinon.

In [87]:
def testLignesEgales(carre):
    """ Renvoie la somme des éléments d'une ligne de la liste 2D carre si toutes les lignes ont la même somme, et -1 sinon """
    if sum(carre[0])==sum(carre[1]) and sum(carre[0])==sum(carre[2]) and sum(carre[0])==sum(carre[3]):
        return sum(carre[0])
    else:
        return -1


print(testLignesEgales(carre_mag))
print(testLignesEgales(carre_pas_mag))

34
-1


In [103]:
sommecarre=0
def testLignesEgales(carre):
    """ Renvoie la somme des éléments d'une ligne de la liste 2D carre si toutes les lignes ont la même somme, et -1 sinon """
    sommecarre=carre[0][0]+carre[0][1]+carre[0][2]+carre[0][3]
    for v in range(4):
        if sommecarre == (carre[v][0]+carre[v][1]+carre[v][2]+carre[v][3]):
            pass
        else:
            return("-1")
    return(sommecarre)


print(testLignesEgales(carre_mag))
print(testLignesEgales(carre_pas_mag))

34
-1


6. Même question mais en testant les colonnes.

In [102]:
sommecarre=0
def testColonnesEgales(carre):
    """ Renvoie la somme des éléments d'une colonne de la liste 2D carre si toutes les colonnes ont la même somme, et -1 sinon """
    sommecarre=carre[0][0]+carre[1][0]+carre[2][0]+carre[3][0]
    for v in range(4):
        if sommecarre == (carre[v][0]+carre[v][1]+carre[v][2]+carre[v][3]):
            pass
        else:
            return("-1")
    return(sommecarre)

print(testColonnesEgales(carre_mag))
print(testColonnesEgales(carre_pas_mag))

34
-1


7. Même question mais en testant les 2 diagonales.

In [111]:
sommecarre=0
def testDiagonalesEgales(carre):
    """ Renvoie la somme des éléments d'une diagonale de la liste 2D carre si les 2 diagonales ont la même somme, et -1 sinon """
    sommecarre=carre[0][0]+carre[1][1]+carre[2][2]+carre[3][3]
    if sommecarre == carre[0][3]+carre[1][2]+carre[2][1]+carre[3][0]:
        return sommecarre
    else:
        return("-1")

print(testDiagonalesEgales(carre_mag))
print(testDiagonalesEgales(carre_pas_mag))
    

34
34


8. Créer une fonction qui teste si un carré est magique. Elle retourne `True` si c'est le cas et `False` sinon.

In [114]:
sommecarre=0
def estCarreMagique(carre):
    sommecarre=carre[0][0]+carre[1][0]+carre[2][0]+carre[3][0]
    """ Renvoie True si c'est un carre magique et False sinon"""
    if testColonnesEgales(carre)==sommecarre and testLignesEgales(carre)==sommecarre and testDiagonalesEgales(carre)==sommecarre:
        return True
    else:
        return False

print(estCarreMagique(carre_mag))
print(estCarreMagique(carre_pas_mag))

True
False


9. Un carré d'ordre $n$ est *normal* s'il contient tous les entiers de 1 à $n^2$. Ecrire une fonction qui teste si un carré est normal (pas nécessairement magique).

In [109]:
def estNormal(carre):
    """ Retourne True si contient toutes les valeurs de 1 à n^2 où n est la taille 
        du carré, et False sinon """
    verifcarre=[]
    n=len(carre)
    for i in range(4):
        for v in range(4):
            verifcarre.append(carre[i][v])
    verifcarre.sort()
    for k in range(1,n^2):
        if k==(verifcarre[k]-1):
            pass
        else:
            return("le carré n'est pas normal")
    return("le carré est normal")

estNormal(carre_mag)

'le carré est normal'